internal/core/component_factory.rs

```diff
@@ -0,0 +1,134 @@
+//! exposed to the C API.
+
+use core::ffi::c_void;
+use core::fmt::Debug;
+
+use alloc::boxed::Box;
+use alloc::rc::Rc;
+
+use crate::{
+    api::ComponentHandle,
+    component::{ComponentRc, ComponentVTable},
+};
+
+/// A `Component` built by a `ComponentFactory`
+#[repr(C)]
+pub struct Component(ComponentRc);
+
+impl<T: ComponentHandle> From<T> for Component
+where
+    T::Inner: vtable::HasStaticVTable<ComponentVTable> + 'static,
+{
+    fn from(value: T) -> Self {
+        Self(vtable::VRc::into_dyn(value.as_weak().inner().upgrade().unwrap()))
+    }
+}
+
+pub(crate) mod component_factory_data {
+    use super::{c_void, Box, Component};
+
+    /// A internal helper needed to implement `ComponentFactory` in different languages.
+    pub(crate) struct ComponentFactoryData {
+        pub build: unsafe extern "C" fn(*mut c_void, *mut c_void),
+        pub data: *mut c_void,
+        pub drop: unsafe extern "C" fn(*mut c_void),
+    }
+
+    #[allow(unsafe_code)]
+    impl ComponentFactoryData {
+        pub fn build(&self) -> Option<Component> {
+            let mut result: Option<Component> = None;
+            unsafe {
+                (self.build)(self.data, &mut result as *mut Option<Component> as *mut c_void)
+            };
+            result
+        }
+    }
+
+    impl Default for ComponentFactoryData {
+        fn default() -> Self {
+            let data = Box::new(RustData { factory: Box::new(|| None) });
+            Self {
+                build: slint_component_factory_builder_call,
+                data: Box::into_raw(data) as *mut c_void,
+                drop: slint_component_factory_builder_drop,
+            }
+        }
+    }
+
+    #[allow(unsafe_code)]
+    impl Drop for ComponentFactoryData {
+        fn drop(&mut self) {
+            unsafe { (self.drop)(self.data) };
+            self.data = core::ptr::null_mut();
+        }
+    }
+
+    pub(crate) struct RustData {
+        pub(crate) factory: Box<dyn Fn() -> Option<Component>>,
+    }
+
+    #[allow(unsafe_code)]
+    pub(crate) unsafe extern "C" fn slint_component_factory_builder_drop(data: *mut c_void) {
+        assert!(!data.is_null());
+        drop(Box::from_raw(data as *mut RustData));
+    }
+
+    // Safety:
+    // `data` must point to the context stored in the `data` field of the
+    // `ComponentFactoryData` this implementation is also a part of!
+    // `result` must point to an `Option<Component>` initialized to `None`!
+    #[allow(unsafe_code)]
+    pub(crate) unsafe extern "C" fn slint_component_factory_builder_call(
+        data: *mut c_void,
+        result: *mut c_void,
+    ) {
+        let result = (result as *mut Option<Component>).as_mut().unwrap();
+        let data = (data as *mut RustData).as_mut().unwrap();
+        *result = (data.factory)();
+    }
+}
+
+/// A Box containing a `ComponentFactory` (or nothing)
+#[derive(Clone, Default)]
+pub struct ComponentFactory(Option<Rc<component_factory_data::ComponentFactoryData>>);
```

Comment:
For now, make it 

```suggestion
pub struct ComponentFactory(Option<Rc<dyn Fn()->Option<Component> + 'static>>);
```

We'll do the C++ part later.
So this patch should almost have no unsafe.